0-module-and-module-source.emu

```diff
@@ -0,0 +1,639 @@
+                2. Else,
+                  1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; a newly created *TypeError* object &raquo;).
+                1. Return ~unused~.
+              1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledClosure_, 0, *""*, &laquo; &raquo;).
+              1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
+                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
+                1. Return ~unused~.
+              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 1, *""*, &laquo; &raquo;).
+              1. Perform PerformPromiseThen(_importHookPromise_, _onFulfilled_, _onRejected_).
+              1. Return ~unused~.
+            </emu-alg>
+
+            <emu-note>
+                ResolveModuleRecordDependency Abstract Operation can not be called more than once for arguments _moduleRecord_ and _specifier_ pair.
+            </emu-note>
+            <emu-note type="editor">
+               Import Reflection Proposal can provide the expansion mechanism to delegate to the host the resolution of a particular dependency, allowing the [[ImportHook]] to return a Module Instance rather than having to open up the ImportHook to support module namespace exotic objects.                  
+            </emu-note>
+        </emu-clause>
+
+        <emu-clause id="sec-hostimportmodulerecorddynamically" type="host-defined abstract operation">
+            <h1>
+              HostImportModuleRecordDynamically (
+                _moduleRecord_: a Module Record,
+                _specifier_: a |ModuleSpecifier| String,
+                _promiseCapability_: a PromiseCapability Record,
+              ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>It performs any necessary setup work in order to make available the module corresponding to module represented by _moduleRecord_. It then performs FinishModuleDynamicImport to finish the dynamic import process.</dd>
+            </dl>
+            <p>An implementation of HostImportModuleRecordDynamically must conform to the following requirements:</p>
+    
+            <ul>
+              <li>
+                It must call ResolveModuleRecordDependency for every (_moduleRecord_, _specifierString_) pair, where _specifierString_ are the entries of _moduleRecord.[[RequestedModules]] for each recursively resolved Cyclic Module Record _moduleRecord_. ResolveModuleRecordDependency must not be called twice with the same (_moduleRecord_, _specifierString_) pair: instead, the result of the first call should be cached and reused as the result of successive resolutions. The resolution for a given pair must not be result of the same Module Record.
```

Comment:
```suggestion
                It must call ResolveModuleRecordDependency for every (_moduleRecord_, _specifierString_) pair, where _specifierString_ are the entries of _moduleRecord_.[[RequestedModules]] for each recursively resolved Cyclic Module Record _moduleRecord_. ResolveModuleRecordDependency must not be called twice with the same (_moduleRecord_, _specifierString_) pair: instead, the result of the first call should be cached and reused as the result of successive resolutions. The resolution for a given pair must not be result of the same Module Record.
```

Reply:
> The resolution for a given pair must not be result of the same Module Record.

Why? A module can import itself.

Reply:
really? cc @erights

Reply:
Verified empirically! In Node.js (using 14 for the moment):

```js
import { name as nom } from './self.mjs';
export const name = 10;
```

Reply:
alright! in that case, I will update the text to remove that wording! that was surprising.